File: packages/workflows/src/incidents/vocabulary.ts
/**
 * EP-019 incident workflow vocabulary (SPEC-018; ADR-026).
 *
 * SPEC-018 locks the canonical terms Incident, Diagnosis, Reproduction,
 * PatchCandidate, Review, Canary, HealthCriterion, Rollback,
 * SkillCandidate, and IncidentMemory. This module adds the
 * EP-019-owned durable workflow vocabulary: the incident remediation
 * workflow kinds and operation kinds the self-healing plane's durable
 * workflows are built from.
 *
 * The canonical lifecycle is OBSERVE -> INCIDENT -> CORRELATE ->
 * DIAGNOSE -> REPRODUCE -> PATCH_PROPOSED -> SANDBOX_VALIDATION ->
 * SECURITY_VALIDATION -> APPROVAL -> STAGED_DEPLOYMENT ->
 * POST_DEPLOY_VERIFICATION -> CLOSED, with explicit terminal/failure
 * states. No state may be collapsed and no model/agent may declare its
 * own fix successful.
 *
 * Every enum here is vocabulary locked: unknown values are rejected at
 * parse time, and a new synonym requires an ADR and schema update
 * (ADR-026).
 */

import { WorkflowContractError } from "../errors.js";

export interface IncidentsVocabulary<T extends string> {
  readonly name: string;
  readonly values: readonly T[];
  is(value: unknown): value is T;
  parse(value: unknown, context?: string): T;
}

export function defineIncidentsVocabulary<T extends string>(
  name: string,
  values: readonly T[],
): IncidentsVocabulary<T> {
  const set = new Set<string>(values);
  return {
    name,
    values,
    is(value: unknown): value is T {
      return typeof value === "string" && set.has(value);
    },
    parse(value: unknown, context = name): T {
      if (typeof value === "string" && set.has(value)) {
        return value as T;
      }
      throw new WorkflowContractError(
        `invalid ${context}: ${JSON.stringify(value)}; expected one of ${values.join(", ")}`,
      );
    },
  };
}

/**
 * Incident remediation workflow kinds (SPEC-018; ADR-026). Each kind is
 * a durable, audited workflow over the self-healing plane. Kinds are
 * EP-019-owned; they are distinct from the EP-006 workflow kinds and
 * are never mixed with them at the registry boundary.
 */
export const incidentWorkflowKind = defineIncidentsVocabulary(
  "IncidentWorkflowKind",
  [
    "INCIDENT_LIFECYCLE",
    "DIAGNOSIS",
    "REPRODUCTION",
    "PATCH_PROPOSAL",
    "REVIEW",
    "CANARY_DEPLOYMENT",
    "ROLLBACK",
  ] as const,
);
export type IncidentWorkflowKind = (typeof incidentWorkflowKind.values)[number];

/**
 * Incident operation kinds: the durable activity-level operations the
 * incident workflows schedule. Each maps to a bounded, idempotent,
 * error-classified activity (SPEC-006 behavior 7). Kinds are
 * EP-019-owned.
 */
export const incidentOperationKind = defineIncidentsVocabulary(
  "IncidentOperationKind",
  [
    "OBSERVE_SIGNAL",
    "CORRELATE_INCIDENT",
    "CREATE_DIAGNOSIS",
    "RUN_REPRODUCTION",
    "GENERATE_PATCH",
    "VALIDATE_SANDBOX",
    "VALIDATE_SECURITY",
    "REQUEST_APPROVAL",
    "STAGE_DEPLOYMENT",
    "VERIFY_POST_DEPLOY",
    "EXECUTE_ROLLBACK",
    "RECORD_MEMORY",
  ] as const,
);
export type IncidentOperationKind =
  (typeof incidentOperationKind.values)[number];

/**
 * Incident lifecycle states (SPEC-018; ADR-026). The exact canonical
 * lifecycle from the EP-019 owner directive, including explicit
 * terminal/failure states. Serializes as SCREAMING_SNAKE_CASE. There is
 * deliberately NO "FIXED"/"REMEDIATED" value: only real observed
 * verification reaches CLOSED.
 */
export const incidentLifecycleState = defineIncidentsVocabulary(
  "IncidentLifecycleState",
  [
    "OBSERVE",
    "INCIDENT",
    "CORRELATE",
    "DIAGNOSE",
    "REPRODUCE",
    "PATCH_PROPOSED",
    "SANDBOX_VALIDATION",
    "SECURITY_VALIDATION",
    "APPROVAL",
    "STAGED_DEPLOYMENT",
    "POST_DEPLOY_VERIFICATION",
    "CLOSED",
    "REJECTED",
    "UNREPRODUCIBLE",
    "VALIDATION_FAILED",
    "SECURITY_FAILED",
    "ROLLED_BACK",
    "BLOCKED",
  ] as const,
);
export type IncidentLifecycleState =
  (typeof incidentLifecycleState.values)[number];

export const INCIDENT_TERMINAL_STATES: readonly IncidentLifecycleState[] = [
  "CLOSED",
  "REJECTED",
  "UNREPRODUCIBLE",
  "VALIDATION_FAILED",
  "SECURITY_FAILED",
  "ROLLED_BACK",
  "BLOCKED",
] as const;

export function isIncidentTerminal(state: IncidentLifecycleState): boolean {
  return (INCIDENT_TERMINAL_STATES as readonly string[]).includes(state);
}

/**
 * Diagnosis confidence (SPEC-018; ADR-026). A model-generated
 * explanation ALWAYS begins as HYPOTHESIS; only reproducible evidence
 * raises it to VALIDATED.
 */
export const diagnosisConfidence = defineIncidentsVocabulary(
  "DiagnosisConfidence",
  ["HYPOTHESIS", "SUPPORTED", "REPRODUCED", "VALIDATED"] as const,
);
export type DiagnosisConfidence = (typeof diagnosisConfidence.values)[number];

/** Independent review verdict (SPEC-018 behavior 4; ADR-026). */
export const reviewVerdict = defineIncidentsVocabulary(
  "IncidentReviewVerdict",
  ["APPROVE", "REJECT", "REQUEST_CHANGES"] as const,
);
export type IncidentReviewVerdict = (typeof reviewVerdict.values)[number];

/** Canary outcome (SPEC-018; ADR-026). */
export const canaryOutcome = defineIncidentsVocabulary("CanaryOutcome", [
  "HEALTHY",
  "PROMOTED",
  "ROLLED_BACK",
  "FAILED",
] as const);
export type CanaryOutcome = (typeof canaryOutcome.values)[number];

/** Rollback outcome (SPEC-018; ADR-026). */
export const rollbackOutcome = defineIncidentsVocabulary("RollbackOutcome", [
  "RESTORED",
  "FAILED",
] as const);
export type RollbackOutcome = (typeof rollbackOutcome.values)[number];

File: crates/nexus-healing/src/vocabulary.rs
//! EP-019 healing vocabulary (SPEC-018; ADR-026).
//!
//! Vocabulary-locked enums for the self-healing engineering loop. The
//! canonical lifecycle is `OBSERVE -> INCIDENT -> CORRELATE -> DIAGNOSE
//! -> REPRODUCE -> PATCH_PROPOSED -> SANDBOX_VALIDATION ->
//! SECURITY_VALIDATION -> APPROVAL -> STAGED_DEPLOYMENT ->
//! POST_DEPLOY_VERIFICATION -> CLOSED` with explicit terminal/failure
//! states (`REJECTED`, `UNREPRODUCIBLE`, `VALIDATION_FAILED`,
//! `SECURITY_FAILED`, `ROLLED_BACK`, `BLOCKED`). Unknown values are
//! rejected at parse time; no free-form strings become domain contracts.
//!
//! DETECTED != DIAGNOSED != REPRODUCED != PATCHED != VERIFIED !=
//! APPROVED != DEPLOYED != REMEDIATED. No state may be collapsed and no
//! model/agent may declare its own fix successful.

use serde::{Deserialize, Serialize};
use std::fmt;

/// Canonical incident lifecycle state (SPEC-018; ADR-026).
///
/// Serializes as SCREAMING_SNAKE_CASE. The states follow the exact
/// canonical ordering in the EP-019 owner directive; `CLOSED` is the
/// only healthy terminal, `REJECTED`/`UNREPRODUCIBLE`/`ROLLED_BACK` are
/// explicit terminal outcomes, and `VALIDATION_FAILED`/`SECURITY_FAILED`/
/// `BLOCKED` are explicit failure terminals. Terminal states never move;
/// resurrection is rejected.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum IncidentState {
    /// A real structured signal was observed (process failure, health
    /// failure, test failure, workflow failure, connector failure,
    /// security event, resource exhaustion, deployment regression).
    Observe,
    /// The signal was accepted as an incident candidate.
    Incident,
    /// The incident was correlated by canonical identifiers (not by raw
    /// error string alone).
    Correlate,
    /// Root-cause work: a diagnosis task exists with confidence.
    Diagnose,
    /// Minimal reproduction attempted (before/after proof).
    Reproduce,
    /// A patch proposal artifact exists.
    PatchProposed,
    /// The patch was validated in an isolated environment.
    SandboxValidation,
    /// Security gates passed for the patch.
    SecurityValidation,
    /// Human/policy approval is required and pending.
    Approval,
    /// The validated patch was deployed to a staged/canary target.
    StagedDeployment,
    /// Post-deploy verification ran against the original reproduction.
    PostDeployVerification,
    /// Incident is remediated: real observed verification closed it.
    Closed,
    /// Explicit terminal: approval denied or patch rejected.
    Rejected,
    /// Explicit terminal: the defect could not be reproduced.
    Unreproducible,
    /// Explicit terminal: sandbox/validation failed.
    ValidationFailed,
    /// Explicit terminal: security validation failed.
    SecurityFailed,
    /// Explicit terminal: rolled back to previous artifact/version.
    RolledBack,
    /// Explicit terminal: blocked with evidence report.
    Blocked,
}

impl IncidentState {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Observe => "OBSERVE",
            Self::Incident => "INCIDENT",
            Self::Correlate => "CORRELATE",
            Self::Diagnose => "DIAGNOSE",
            Self::Reproduce => "REPRODUCE",
            Self::PatchProposed => "PATCH_PROPOSED",
            Self::SandboxValidation => "SANDBOX_VALIDATION",
            Self::SecurityValidation => "SECURITY_VALIDATION",
            Self::Approval => "APPROVAL",
            Self::StagedDeployment => "STAGED_DEPLOYMENT",
            Self::PostDeployVerification => "POST_DEPLOY_VERIFICATION",
            Self::Closed => "CLOSED",
            Self::Rejected => "REJECTED",
            Self::Unreproducible => "UNREPRODUCIBLE",
            Self::ValidationFailed => "VALIDATION_FAILED",
            Self::SecurityFailed => "SECURITY_FAILED",
            Self::RolledBack => "ROLLED_BACK",
            Self::Blocked => "BLOCKED",
        }
    }

    pub const fn is_terminal(self) -> bool {
        matches!(
            self,
            Self::Closed
                | Self::Rejected
                | Self::Unreproducible
                | Self::ValidationFailed
                | Self::SecurityFailed
                | Self::RolledBack
                | Self::Blocked
        )
    }

    pub const fn is_healthy_terminal(self) -> bool {
        matches!(self, Self::Closed)
    }

    pub const ALL: [IncidentState; 18] = [
        Self::Observe,
        Self::Incident,
        Self::Correlate,
        Self::Diagnose,
        Self::Reproduce,
        Self::PatchProposed,
        Self::SandboxValidation,
        Self::SecurityValidation,
        Self::Approval,
        Self::StagedDeployment,
        Self::PostDeployVerification,
        Self::Closed,
        Self::Rejected,
        Self::Unreproducible,
        Self::ValidationFailed,
        Self::SecurityFailed,
        Self::RolledBack,
        Self::Blocked,
    ];
}

impl fmt::Display for IncidentState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl std::str::FromStr for IncidentState {
    type Err = super::error::HealingError;
    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "OBSERVE" => Ok(Self::Observe),
            "INCIDENT" => Ok(Self::Incident),
            "CORRELATE" => Ok(Self::Correlate),
            "DIAGNOSE" => Ok(Self::Diagnose),
            "REPRODUCE" => Ok(Self::Reproduce),
            "PATCH_PROPOSED" => Ok(Self::PatchProposed),
            "SANDBOX_VALIDATION" => Ok(Self::SandboxValidation),
            "SECURITY_VALIDATION" => Ok(Self::SecurityValidation),
            "APPROVAL" => Ok(Self::Approval),
            "STAGED_DEPLOYMENT" => Ok(Self::StagedDeployment),
            "POST_DEPLOY_VERIFICATION" => Ok(Self::PostDeployVerification),
            "CLOSED" => Ok(Self::Closed),
            "REJECTED" => Ok(Self::Rejected),
            "UNREPRODUCIBLE" => Ok(Self::Unreproducible),
            "VALIDATION_FAILED" => Ok(Self::ValidationFailed),
            "SECURITY_FAILED" => Ok(Self::SecurityFailed),
            "ROLLED_BACK" => Ok(Self::RolledBack),
            "BLOCKED" => Ok(Self::Blocked),
            other => Err(super::error::HealingError::vocabulary(
                "IncidentState",
                other,
            )),
        }
    }
}

/// Diagnosis confidence (SPEC-018; ADR-026).
///
/// A model-generated explanation ALWAYS begins as `HYPOTHESIS` and only
/// becomes `VALIDATED` after reproducible evidence supports it. A
/// hypothesis is never root cause by assertion.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DiagnosisConfidence {
    /// Model/agent explanation, no reproducible evidence yet.
    Hypothesis,
    /// Correlated evidence supports the explanation but it is not yet
    /// reproduced.
    Supported,
    /// A minimal reproduction exhibits the failure.
    Reproduced,
    /// Reproducible evidence plus verification support the explanation.
    Validated,
}

impl DiagnosisConfidence {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Hypothesis => "HYPOTHESIS",
            Self::Supported => "SUPPORTED",
            Self::Reproduced => "REPRODUCED",
            Self::Validated => "VALIDATED",
        }
    }

    pub const fn is_authoritative(self) -> bool {
        matches!(self, Self::Validated)
    }

    pub const ALL: [DiagnosisConfidence; 4] = [
        Self::Hypothesis,
        Self::Supported,
        Self::Reproduced,
        Self::Validated,
    ];
}

impl fmt::Display for DiagnosisConfidence {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

impl std::str::FromStr for DiagnosisConfidence {
    type Err = super::error::HealingError;
    fn from_str(value: &str) -> Result<Self, Self::Err> {
        match value {
            "HYPOTHESIS" => Ok(Self::Hypothesis),
            "SUPPORTED" => Ok(Self::Supported),
            "REPRODUCED" => Ok(Self::Reproduced),
            "VALIDATED" => Ok(Self::Validated),
            other => Err(super::error::HealingError::vocabulary(
                "DiagnosisConfidence",
                other,
            )),
        }
    }
}
